File: src/AStarSearch.java
import java.util.*;

public class AStarSearch {

    public static Result search(List<Crop> initialCrops, List<Fertilizer> fertilizers, int maxCycles) {
        PriorityQueue<Node> openSet = new PriorityQueue<>(Comparator.comparingInt(n -> n.fCost));
        Set<String> visited = new HashSet<>();

        List<ActiveFertilizer> activeFerts = new ArrayList<>();
        Node start = new Node(copyCrops(initialCrops), new ArrayList<>(), 0, 0, activeFerts);
        openSet.add(start);

        Node best = null;

        while (!openSet.isEmpty()) {
            Node current = openSet.poll();

            String stateKey = generateStateKey(current.crops, current.activeFertilizers);
            if (visited.contains(stateKey)) continue;
            visited.add(stateKey);

            if (current.plan.size() == maxCycles) {
                if (best == null || totalYield(current.crops) > totalYield(best.crops)) {
                    best = current;
                }
                continue;
            }

            for (Fertilizer fert : fertilizers) {
                List<Crop> nextCrops = copyCrops(current.crops);
                List<ActiveFertilizer> nextActive = copyActiveFerts(current.activeFertilizers);

                if (!fert.name.equals("None")) {
                    applyFertilizer(nextCrops, fert, nextActive);
                }

                simulateCycle(nextCrops, nextActive);
                List<Fertilizer> nextPlan = new ArrayList<>(current.plan);
                nextPlan.add(fert);

                int cost = current.totalCost + (fert.name.equals("None") ? 0 : fert.cost); // Increase cost only if fertilizer is not "None"

                // Estimate heuristics
                int yieldScore = totalYield(nextCrops);
                int affectedTypes = (int) nextCrops.stream()
                        .map(c -> c.type)
                        .filter(type -> fert.affectsType(type))
                        .distinct()
                        .count();

                // int heuristic = -(yieldScore * 1000 + affectedTypes * 900); // small bonus for diversity
                // int heuristic = -(yieldScore * affectedTypes * 1000); // multiply bonus for diversity
                int heuristic = -(yieldScore * 1000 + (int)Math.pow(affectedTypes, 2) * 1500); // exponentially increase bonus for diversity

                int fCost = cost + heuristic;

                openSet.add(new Node(nextCrops, nextPlan, cost, fCost, nextActive));
            }
        }

        return best == null ? null : new Result(best.plan, best.crops, best.totalCost);
    }

    private static void applyFertilizer(List<Crop> crops, Fertilizer fert, List<ActiveFertilizer> active) {
        ActiveFertilizer af = new ActiveFertilizer(fert, fert.duration);
        active.add(af);

        if (fert.effect.equals("Soft Soil")) {
            for (ActiveFertilizer a : active) {
                if (a != af) {
                    a.remainingDuration += 5;
                }
            }
        }
    }

    private static void simulateCycle(List<Crop> crops, List<ActiveFertilizer> activeFerts) {
        for (Crop crop : crops) {
            int growthTime = crop.baseGrowthTime;
            int yieldBonus = 0;

            for (ActiveFertilizer af : activeFerts) {
                if (af.remainingDuration <= 0) continue;

                String effect = af.fertilizer.effect;

                if (effect.equals("Catalyst")) {
                    growthTime = Math.max(1, growthTime - 1); // To avoid growthTime being zero or negative
                }
                if (effect.equals("AncientCatalyst")) {
                    growthTime = 1;
                }
                if (effect.equals("BoostPlant") && crop.type.equals("Plant")) {
                    yieldBonus += crop.baseYield + (crop.baseYield / 2);
                }
                if (effect.equals("BoostInsect") && crop.type.equals("Insect")) {
                    yieldBonus += crop.baseYield + (crop.baseYield / 2);
                }
                if (effect.equals("BoostMushroom") && crop.type.equals("Mushroom")) {
                    yieldBonus += crop.baseYield + (crop.baseYield / 2);
                }
            }

            crop.currentGrowth++;
            if (crop.currentGrowth >= growthTime) {
                crop.yield += crop.baseYield + yieldBonus;
                crop.currentGrowth = 0;
            }
        }

        // Decrement durations
        activeFerts.removeIf(f -> --f.remainingDuration <= 0);
    }

    private static List<Crop> copyCrops(List<Crop> crops) {
        List<Crop> copy = new ArrayList<>();
        for (Crop c : crops) {
            copy.add(c.copy());
        }
        return copy;
    }

    private static List<ActiveFertilizer> copyActiveFerts(List<ActiveFertilizer> list) {
        List<ActiveFertilizer> copy = new ArrayList<>();
        for (ActiveFertilizer f : list) {
            copy.add(f.copy());
        }
        return copy;
    }

    private static int totalYield(List<Crop> crops) {
        return crops.stream().mapToInt(c -> c.yield).sum();
    }

    private static String generateStateKey(List<Crop> crops, List<ActiveFertilizer> activeFerts) {
        StringBuilder sb = new StringBuilder();
        for (Crop c : crops) {
            sb.append(c.name).append(":").append(c.currentGrowth).append(":").append(c.yield).append(";");
        }
        for (ActiveFertilizer af : activeFerts) {
            sb.append(af.fertilizer.name).append(":").append(af.remainingDuration).append(";");
        }
        return sb.toString();
    }

    public static class Result {
        public final List<Fertilizer> plan;
        public final List<Crop> finalCrops;
        public final int totalCost;

        public Result(List<Fertilizer> plan, List<Crop> finalCrops, int totalCost) {
            this.plan = plan;
            this.finalCrops = finalCrops;
            this.totalCost = totalCost;
        }
    }

    private static class Node {
        public final List<Crop> crops;
        public final List<Fertilizer> plan;
        public final int totalCost;
        public final int fCost;
        public final List<ActiveFertilizer> activeFertilizers;

        public Node(List<Crop> crops, List<Fertilizer> plan, int totalCost, int fCost, List<ActiveFertilizer> activeFertilizers) {
            this.crops = crops;
            this.plan = plan;
            this.totalCost = totalCost;
            this.fCost = fCost;
            this.activeFertilizers = activeFertilizers;
        }
    }
}

File: src/AStarSearch.py
import heapq
from dataclasses import dataclass, field
from typing import List, Set, Tuple


@dataclass
class Fertilizer:
    name: str
    duration: int
    cost: int
    effect: str

    def affects_type(self, crop_type: str) -> bool:
        if self.effect == "BoostPlant" and crop_type == "Plant":
            return True
        if self.effect == "BoostInsect" and crop_type == "Insect":
            return True
        if self.effect == "BoostMushroom" and crop_type == "Mushroom":
            return True
        return False


@dataclass
class Crop:
    name: str
    baseYield: int
    baseGrowthTime: int
    type: str
    currentGrowth: int = 0
    yield_: int = 0

    def copy(self):
        return Crop(self.name, self.baseYield, self.baseGrowthTime, self.type, self.currentGrowth, self.yield_)


@dataclass
class ActiveFertilizer:
    fertilizer: Fertilizer
    remainingDuration: int

    def copy(self):
        return ActiveFertilizer(self.fertilizer, self.remainingDuration)


@dataclass(order=True)
class Node:
    fCost: int
    crops: List[Crop] = field(compare=False)
    plan: List[Fertilizer] = field(compare=False)
    totalCost: int = field(compare=False)
    activeFertilizers: List[ActiveFertilizer] = field(compare=False)


@dataclass
class Result:
    plan: List[Fertilizer]
    finalCrops: List[Crop]
    totalCost: int


class AStarSearch:

    @staticmethod
    def search(initialCrops: List[Crop], fertilizers: List[Fertilizer], maxCycles: int) -> Result:
        openSet: List[Node] = []
        visited: Set[str] = set()

        start = Node(
            fCost=0,
            crops=AStarSearch.copyCrops(initialCrops),
            plan=[],
            totalCost=0,
            activeFertilizers=[]
        )
        heapq.heappush(openSet, start)

        best = None

        while openSet:
            current = heapq.heappop(openSet)
            stateKey = AStarSearch.generateStateKey(current.crops, current.activeFertilizers)

            if stateKey in visited:
                continue
            visited.add(stateKey)

            if len(current.plan) == maxCycles:
                if best is None or AStarSearch.totalYield(current.crops) > AStarSearch.totalYield(best.crops):
                    best = current
                continue

            for fert in fertilizers:
                nextActive = AStarSearch.copyActiveFerts(current.activeFertilizers)

                alreadyActive = any(
                    a.fertilizer.name == fert.name and a.remainingDuration > 0
                    for a in nextActive
                )
                if fert.name != "None" and alreadyActive:
                    continue

                nextCrops = AStarSearch.copyCrops(current.crops)

                if fert.name != "None":
                    AStarSearch.applyFertilizer(nextCrops, fert, nextActive)

                AStarSearch.simulateCycle(nextCrops, nextActive)

                nextPlan = current.plan + [fert]
                cost = current.totalCost + (0 if fert.name == "None" else fert.cost)

                yieldScore = AStarSearch.totalYield(nextCrops)
                affectedTypes = len({
                    c.type for c in nextCrops if fert.affects_type(c.type)
                })

                heuristic = -(yieldScore + (affectedTypes ** 2) * 1000)
                fCost = cost + heuristic

                heapq.heappush(openSet, Node(fCost, nextCrops, nextPlan, cost, nextActive))

        return Result(best.plan, best.crops, best.totalCost) if best else None

    @staticmethod
    def applyFertilizer(crops: List[Crop], fert: Fertilizer, active: List[ActiveFertilizer]):
        af = ActiveFertilizer(fert, fert.duration)
        active.append(af)

        if fert.effect == "Soft Soil":
            for a in active:
                if a != af:
                    a.remainingDuration += 5

    @staticmethod
    def simulateCycle(crops: List[Crop], activeFerts: List[ActiveFertilizer]):
        for crop in crops:
            growthTime = crop.baseGrowthTime
            yieldBonus = 0

            for af in activeFerts:
                if af.remainingDuration <= 0:
                    continue

                effect = af.fertilizer.effect
                if effect == "Catalyst":
                    growthTime = max(1, growthTime - 1)
                if effect == "AncientCatalyst":
                    growthTime = 1
                if effect == "BoostPlant" and crop.type == "Plant":
                    yieldBonus += crop.baseYield // 2
                if effect == "BoostInsect" and crop.type == "Insect":
                    yieldBonus += crop.baseYield // 2
                if effect == "BoostMushroom" and crop.type == "Mushroom":
                    yieldBonus += crop.baseYield // 2

            crop.currentGrowth += 1
            if crop.currentGrowth >= growthTime:
                crop.yield_ += crop.baseYield + yieldBonus
                crop.currentGrowth = 0

        # activeFerts[:] = [f for f in activeFerts if (f.remainingDuration := f.remainingDuration - 1) > 0]
        new_active_ferts = []
        for f in activeFerts:
            f.remainingDuration -= 1
            if f.remainingDuration > 0:
                new_active_ferts.append(f)
        activeFerts[:] = new_active_ferts

    @staticmethod
    def copyCrops(crops: List[Crop]) -> List[Crop]:
        return [c.copy() for c in crops]

    @staticmethod
    def copyActiveFerts(activeFerts: List[ActiveFertilizer]) -> List[ActiveFertilizer]:
        return [a.copy() for a in activeFerts]

    @staticmethod
    def totalYield(crops: List[Crop]) -> int:
        return sum(c.yield_ for c in crops)

    @staticmethod
    def generateStateKey(crops: List[Crop], activeFerts: List[ActiveFertilizer]) -> str:
        crop_key = ';'.join(f"{c.name}:{c.currentGrowth}:{c.yield_}" for c in crops)
        fert_key = ';'.join(f"{af.fertilizer.name}:{af.remainingDuration}" for af in activeFerts)
        return crop_key + '|' + fert_key
